### PR TITLE
fix(reserve): optimizing the cleanup worker

### PR DESCRIPTION
**Pull request by istae**
### Checklist

- [ ] I have read the [coding guide](https://github.com/ethersphere/bee/blob/master/CODING.md).
- [ ] My change requires a documentation update, and I have done it.
- [ ] I have added tests to cover my changes.
- [ ] I have filled out the description and linked the related issues.

### CODE REVIEW

File: pkg/storer/reserve.go [hunk 1]
```diff
@@ -276,21 +281,29 @@ func (db *DB) removeChunk(ctx context.Context, address swarm.Address, batchID []
...
 	dur := captureDuration(time.Now())
+	count := 0
 	defer func() {
 		db.metrics.MethodCallsDuration.WithLabelValues("reserve", "cleanup").Observe(dur())
+		db.metrics.ReserveSize.Set(float64(db.reserve.Size()))
+		db.logger.Info("cleanup finished", "total", count, "duration", dur())
 	}()
 
-	return db.reserve.IterateChunksItems(db.repo, 0, func(ci reserve.ChunkItem) (bool, error) {
+	ids := map[string]struct{}{}
```
Comment (ldeffenb) [inline pkg/storer/reserve.go:291]: Anxious to see if this improves (reduces) the elapsed time of reserveCleanup!

File: pkg/storer/reserve.go [hunk 2]
```diff
@@ -276,21 +281,29 @@ func (db *DB) removeChunk(ctx context.Context, address swarm.Address, batchID []
 
 func (db *DB) reserveCleanup(ctx context.Context) error {
 	dur := captureDuration(time.Now())
+	count := 0
 	defer func() {
 		db.metrics.MethodCallsDuration.WithLabelValues("reserve", "cleanup").Observe(dur())
+		db.metrics.ReserveSize.Set(float64(db.reserve.Size()))
+		db.logger.Info("cleanup finished", "total", count, "duration", dur())
```
Comment (ldeffenb) [inline pkg/storer/reserve.go:288]: Would "removed" be better than "total" the latter of which might imply that it's the total size of the reserve?

**Review by ldeffenb — CHANGES_REQUESTED** [review]

File: pkg/storer/reserve.go [hunk 3]
```diff
@@ -276,21 +281,29 @@ func (db *DB) removeChunk(ctx context.Context, address swarm.Address, batchID []
 
 func (db *DB) reserveCleanup(ctx context.Context) error {
 	dur := captureDuration(time.Now())
+	removed := 0
 	defer func() {
 		db.metrics.MethodCallsDuration.WithLabelValues("reserve", "cleanup").Observe(dur())
+		db.metrics.ReserveSize.Set(float64(db.reserve.Size()))
```
Comment (aloknerurkar) [inline pkg/storer/reserve.go:287]: what about the Reserve.Size ?
We set the metric but we are not adjusting the size of the reserve. Also there is no easy way to do it as we are iterating with bin 0. So there could be chunks which are not accounted for in the reserve size. We will need to do a complete recalculation after the cleanup is done I think.

File: pkg/storer/reserve.go [hunk 4]
```diff
@@ -276,21 +281,29 @@ func (db *DB) removeChunk(ctx context.Context, address swarm.Address, batchID []
...
+		db.logger.Info("cleanup finished", "removed", removed, "duration", dur())
```
Comment (aloknerurkar) [inline pkg/storer/reserve.go:288]: We should also add a metric for this. If this happens a lot in 6 hrs we have a problem.

File: pkg/storer/reserve.go (hunk 3 above)
Comment (aloknerurkar) [reply pkg/storer/reserve.go:287]: This is happening right now. After restart the reserve values drop.

Screenshot 2023-06-28 at 11 54 01 PM

Comment (ldeffenb) [reply pkg/storer/reserve.go:287]: > what about the Reserve.Size ?

He's right.  I missed the calls to db.reserve.AddSize(1) and db.reserve.AddSize(-evicted) in ReservePutter and evictBatch above.

Comment (istae) [reply pkg/storer/reserve.go:287]: nice catch

**Review by ldeffenb — APPROVED** [review]
It all looks correct to me.   I've got an idea on lazy-population of the ids[] which can go into a different PR at some point in the future.

File: pkg/storer/reserve.go [hunk 5]
```diff
@@ -276,21 +281,39 @@ func (db *DB) removeChunk(ctx context.Context, address swarm.Address, batchID []
...
 	dur := captureDuration(time.Now())
+	removed := 0
 	defer func() {
 		db.metrics.MethodCallsDuration.WithLabelValues("reserve", "cleanup").Observe(dur())
-	}()
+		db.metrics.ReserveCleanup.Add(float64(removed))
+		db.logger.Info("cleanup finished", "removed", removed, "duration", dur())
 
-	return db.reserve.IterateChunksItems(db.repo, 0, func(ci reserve.ChunkItem) (bool, error) {
+		db.lock.Lock(reserveUpdateLockKey)
```
Comment (aloknerurkar) [inline pkg/storer/reserve.go:290]: We should only do this `if removed > 0`

File: pkg/storer/reserve.go [hunk 6]
```diff
@@ -276,21 +281,39 @@ func (db *DB) removeChunk(ctx context.Context, address swarm.Address, batchID []
...
+		return false, nil
+	})
+	if err != nil {
+		return err
+	}
+
+	return db.reserve.IterateChunksItems(db.repo, 0, func(ci reserve.ChunkItem) (bool, error) {
+		if _, ok := ids[string(ci.BatchID)]; !ok {
+			removed++
+			db.logger.Debug("cleanup expired batch", "batch_id", hex.EncodeToString(ci.BatchID))
```
Comment (mrekucci) [inline pkg/storer/reserve.go:314]: Logging might slow the iteration.

**Review by mrekucci — APPROVED** [review]
Apart from the comments, LGTM.

File: pkg/storer/reserve.go (hunk 6 above)
Comment (istae) [reply pkg/storer/reserve.go:314]: we should only catch a few chunks here :)

**Review by aloknerurkar — APPROVED** [review]

**Pull request merged**